### PR TITLE
fix(agw): c++ migration of nas files

### PR DESCRIPTION
**Pull request by pruthvihebbani**
fix(agw): c++ migration of nas files

## Summary

Migrated all the files in the following folders to cpp/hpp and fixed all the compilation errors and warnings:
1. nas
2. nas/api
3. nas/emm
4. nas/util

## Test Plan

Verified s1ap integ_tests and unit test

### CODE REVIEW

File: lte/gateway/c/core/BUILD.bazel [hunk 1]
```diff
@@ -692,13 +692,15 @@ MME_HDRS = [
...
-    "oai/tasks/nas/emm/emm_proc.h",
+    "oai/tasks/nas/api/mme/mme_api.hpp",
+    "oai/tasks/nas/api/network/nas_message.hpp",
+    "oai/tasks/nas/emm/EmmCommon.hpp",
+    "oai/tasks/nas/emm/LowerLayer.hpp",
+    "oai/tasks/nas/emm/emm_data.hpp",
+    "oai/tasks/nas/emm/emm_headers.hpp",
+    "oai/tasks/nas/emm/emm_main.hpp",
+    "oai/tasks/nas/emm/emm_proc.hpp",
+    "oai/tasks/nas/emm/emm_headers.hpp",
```
Comment (LKreutzer) [inline lte/gateway/c/core/BUILD.bazel]: `"oai/tasks/nas/emm/emm_headers.hpp",` appears twice. In lines 700 and 703.

**Review by LKreutzer — CHANGES_REQUESTED** [review]

**Review by LKreutzer — CHANGES_REQUESTED** [review]
The following file is not covered by a `BUILD.bazel` file:
`./lte/gateway/c/core/oai/tasks/nas/nas_network.cpp`

File: lte/gateway/c/core/BUILD.bazel (hunk 1 above)
Comment (pruthvihebbani) [reply lte/gateway/c/core/BUILD.bazel]: Resolved

**Review by LKreutzer — CHANGES_REQUESTED** [review]
The bazel build is failing due to a casting error in the file `socket.cpp`.
Please replace the `socket_get_fd` function with the following and add the `#include <iostream>` header to the file.
```
status_code_e socket_get_fd(const void* id) {
  if (id) {
    int socket_file_descriptor = const_cast<socket_id_t*>(reinterpret_cast<const socket_id_t*>(id))->fd;
    return static_cast<status_code_e>(socket_file_descriptor);
  }

  return RETURNerror;
}
```

Please address the other type conversion/linter warnings.

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_defs.h [hunk 2]
```diff
@@ -35,7 +35,13 @@
...
 #include "lte/gateway/c/core/oai/tasks/mme_app/mme_app_sgs_fsm.h"
-#include "lte/gateway/c/core/oai/tasks/nas/emm/emm_proc.h"
+#ifdef __cplusplus
+extern "C" {
+#endif
+#include "lte/gateway/c/core/oai/tasks/nas/emm/emm_proc.hpp"
+#ifdef __cplusplus
+}
+#endif
 #include <czmq.h>
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_defs.h]: Move this line above, as this is part of zmq library

File: lte/gateway/c/core/oai/tasks/nas/api/mme/mme_api.cpp [hunk 3]
```diff
@@ -32,25 +32,32 @@
...
+extern "C" {
+#endif
 #include "lte/gateway/c/core/common/assertions.h"
 #include "lte/gateway/c/core/common/common_defs.h"
 #include "lte/gateway/c/core/oai/common/common_types.h"
 #include "lte/gateway/c/core/oai/common/conversions.h"
 #include "lte/gateway/c/core/oai/common/log.h"
+#ifdef __cplusplus
+}
+#endif
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/api/mme/mme_api.cpp:53]: Leave the space

File: lte/gateway/c/core/oai/tasks/nas/emm/Attach.cpp [hunk 4]
```diff
@@ -1644,10 +1659,7 @@ static int emm_send_attach_accept(emm_context_t* emm_context) {
       if (!IS_EMM_CTXT_PRESENT_GUTI(emm_context)) {
         // Sure it is an unknown GUTI in this MME
         guti_t old_guti = emm_context->_old_guti;
-        guti_t guti = {.gummei.plmn = {0},
-                       .gummei.mme_gid = 0,
-                       .gummei.mme_code = 0,
-                       .m_tmsi = INVALID_M_TMSI};
+        guti_t guti = {};
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/Attach.cpp:1665]: The value of m_tmsi needs to be set to INVALID_M_TMSI

File: lte/gateway/c/core/oai/tasks/nas/emm/Authentication.cpp [hunk 5]
```diff
@@ -20,12 +20,20 @@
 #include <string.h>
 #include <stdlib.h>
 
+#ifdef __cplusplus
+extern "C" {
+#endif
 #include "lte/gateway/c/core/common/assertions.h"
-#include "lte/gateway/c/core/common/common_defs.h"
 #include "lte/gateway/c/core/common/dynamic_memory_check.h"
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/Authentication.cpp]: Place the dynamic_memory_check.h outside of extern C.

File: lte/gateway/c/core/oai/tasks/nas/emm/EmmCommon.cpp [hunk 6]
```diff
@@ -41,18 +41,24 @@
...
 #include "lte/gateway/c/core/common/common_defs.h"
-#include "lte/gateway/c/core/common/dynamic_memory_check.h"
-#include "lte/gateway/c/core/oai/common/log.h"
 #include "lte/gateway/c/core/oai/include/mme_config.h"
-#include "lte/gateway/c/core/oai/tasks/nas/emm/EmmCommon.h"
-
+#ifdef __cplusplus
+}
+#endif
+#include "lte/gateway/c/core/common/dynamic_memory_check.h"
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/EmmCommon.cpp:61]: Space

File: lte/gateway/c/core/oai/tasks/nas/emm/LowerLayer.cpp [hunk 7]
```diff
@@ -20,21 +20,28 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
 #include "lte/gateway/c/core/common/common_defs.h"
 #include "lte/gateway/c/core/oai/common/common_types.h"
 #include "lte/gateway/c/core/oai/common/log.h"
+#include "lte/gateway/c/core/oai/lib/bstr/bstrlib.h"
+#ifdef __cplusplus
+}
+#endif
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/LowerLayer.cpp:34]: Space

File: lte/gateway/c/core/oai/tasks/nas/emm/emm_main.cpp [hunk 8]
```diff
@@ -17,12 +17,18 @@
...
 
+#ifdef __cplusplus
+extern "C" {
+#endif
 #include "lte/gateway/c/core/common/assertions.h"
 #include "lte/gateway/c/core/common/common_defs.h"
 #include "lte/gateway/c/core/oai/common/log.h"
+#ifdef __cplusplus
+}
+#endif
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/emm_main.cpp:28]: Space

File: lte/gateway/c/core/oai/tasks/nas/emm/emm_proc.hpp [hunk 9]
```diff
@@ -40,9 +40,9 @@ Description Defines the EPS Mobility Management procedures executed at
 #include "lte/gateway/c/core/oai/include/nas/commonDef.h"
 #include "lte/gateway/c/core/oai/lib/bstr/bstrlib.h"
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/emm_proc.hpp:44]: Include it within extern C

File: lte/gateway/c/core/oai/tasks/nas/emm/sap/emm_as.cpp [hunk 10]
```diff
@@ -21,11 +21,19 @@
 #include <stdlib.h>
 #include <netinet/in.h>
 
-#include "lte/gateway/c/core/common/common_defs.h"
 #include "lte/gateway/c/core/common/dynamic_memory_check.h"
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/sap/emm_as.cpp]: I think this inclusion comes after extern C

File: lte/gateway/c/core/oai/tasks/nas/emm/sap/emm_as.cpp [hunk 11]
```diff
@@ -21,11 +21,19 @@
...
+#include "lte/gateway/c/core/common/common_defs.h"
+#include "lte/gateway/c/core/oai/common/log.h"
 #include "lte/gateway/c/core/oai/common/common_types.h"
 #include "lte/gateway/c/core/oai/common/conversions.h"
-#include "lte/gateway/c/core/oai/common/log.h"
+#include "lte/gateway/c/core/oai/lib/bstr/bstrlib.h"
+#include "lte/gateway/c/core/oai/tasks/mme_app/mme_app_itti_messaging.h"
+#ifdef __cplusplus
+}
+#endif
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/sap/emm_as.cpp:35]: General comment to leave space

File: lte/gateway/c/core/oai/tasks/nas/emm/sap/emm_as.cpp [hunk 12]
```diff
@@ -282,17 +291,16 @@ status_code_e emm_as_send(emm_as_t* msg) {
...
-                       int* emm_cause,
-                       nas_message_decode_status_t* decode_status) {
+static status_code_e emm_as_recv(mme_ue_s1ap_id_t ue_id,
+                                 tai_t const* originating_tai,
+                                 ecgi_t const* originating_ecgi, bstring msg,
+                                 size_t len, int* emm_cause,
+                                 nas_message_decode_status_t* decode_status) {
   OAILOG_FUNC_IN(LOG_NAS_EMM);
   nas_message_decode_status_t local_decode_status = {0};
   int decoder_rc = RETURNok;
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/sap/emm_as.cpp]: Initialize decode_Rc to 0

**Review by LKreutzer — CHANGES_REQUESTED** [review]
After extended debugging with @nstng we found that in the current state the MME build is failing if gcc has an optimisation level higher than `-O0`. 

The optimisation is turned on in the bazel build with the `--config=production` and this is why the CI job is failing. 

In the Make build the optimisation, for `BUILD_TYPE="RelWithDebInfo` (which is not run in CI), is set to `-O1` for C files but not for C++ files - see `lte/gateway/c/core/oai/CMakeLists.txt`. Compare
- line 51 `set(CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS} -g -DMALLOC_CHECK_=3  -O1")`
- line 62 `set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} -fsanitize=leak -fno-omit-frame-pointer")`

If we set line 62 to `set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} -fsanitize=leak -fno-omit-frame-pointer -O1")` then the Make build is also failing. To reproduce make this change and run `make build_oai BUILD_TYPE="RelWithDebInfo"`.

@pruthvihebbani could you please make sure that `bazel build --config=production //lte/gateway/c/core:agw_of` is working. In this case I expect that the above Make command with the optimisation for C++ turned on will also work.

@ssanadhya Do you know why the `-O1` flag in `lte/gateway/c/core/oai/CMakeLists.txt` for `BUILD_TYPE="RelWithDebInfo` is only set for C files but not for C++ files?

**Review by LKreutzer — APPROVED** [review]
Bazel changes lgtm

File: lte/gateway/c/core/oai/include/mme_config.h [hunk 13]
```diff
@@ -45,7 +45,13 @@
...
 #include "lte/gateway/c/core/oai/lib/3gpp/3gpp_24.008.h"
+#ifdef __cplusplus
+extern "C" {
+#endif
 #include "lte/gateway/c/core/oai/common/log.h"
+#ifdef __cplusplus
+}
+#endif
 #include "lte/gateway/c/core/oai/include/service303.hpp"
 #include "lte/gateway/c/core/oai/lib/hashtable/hashtable.h"
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/include/mme_config.h]: Move the header files of hash library under extern "C" and leave space before after the extern "C" statements

File: lte/gateway/c/core/oai/lib/secu/secu_defs.h [hunk 14]
```diff
@@ -31,9 +31,14 @@ void kdf(const uint8_t* key, const unsigned key_len, uint8_t* s,
...
 
+#ifdef __cplusplus
+extern "C" {
+#endif
 int derive_key_nas(algorithm_type_dist_t nas_alg_type, uint8_t nas_enc_alg_id,
                    const uint8_t* kasme_32, uint8_t* knas);
-
+#ifdef __cplusplus
+}
+#endif
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/lib/secu/secu_defs.h:41]: Space

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_defs.h [hunk 15]
```diff
@@ -30,13 +30,20 @@
 #ifndef FILE_MME_APP_DEFS_SEEN
 #define FILE_MME_APP_DEFS_SEEN
 
-#include "lte/gateway/c/core/oai/lib/itti/intertask_interface.h"
+#include <czmq.h>
 
+#include "lte/gateway/c/core/oai/lib/itti/intertask_interface.h"
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_defs.h]: Move  under extern "C". General comment is to move hash library related header files under extern "C".

File: lte/gateway/c/core/oai/tasks/nas/api/mme/mme_api.hpp [hunk 16]
```diff
@@ -156,32 +156,40 @@ status_code_e mme_api_unsubscribe(bstring apn);
 }
 #endif
 
-status_code_e mme_api_notify_imsi(const mme_ue_s1ap_id_t id,
-                                  const imsi64_t imsi64);
-
 status_code_e mme_api_notify_new_guti(const mme_ue_s1ap_id_t ueid,
                                       guti_t* const guti);
 
+/*TODO: These declarations are temporarily moved to emm_headers.hpp file to
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/api/mme/mme_api.hpp:162]: Could you mention the issue number, https://github.com/magma/magma/issues/13096
The care should be taken while handling this issue

Good to mention this issue id on all occurrences of this comment

File: lte/gateway/c/core/oai/tasks/nas/emm/Identification.cpp [hunk 17]
```diff
@@ -19,11 +19,17 @@
...
+extern "C" {
+#endif
 #include "lte/gateway/c/core/common/assertions.h"
 #include "lte/gateway/c/core/common/common_defs.h"
 #include "lte/gateway/c/core/oai/common/common_types.h"
 #include "lte/gateway/c/core/oai/common/conversions.h"
 #include "lte/gateway/c/core/oai/common/log.h"
+#ifdef __cplusplus
+}
+#endif
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/Identification.cpp:32]: Space

File: lte/gateway/c/core/oai/tasks/nas/emm/SecurityModeControl.cpp [hunk 18]
```diff
@@ -46,14 +46,20 @@
...
+#endif
 #include "lte/gateway/c/core/common/assertions.h"
 #include "lte/gateway/c/core/common/common_defs.h"
 #include "lte/gateway/c/core/oai/common/common_types.h"
 #include "lte/gateway/c/core/oai/common/conversions.h"
 #include "lte/gateway/c/core/oai/common/log.h"
+#include "lte/gateway/c/core/oai/include/mme_app_ue_context.h"
+#ifdef __cplusplus
+}
+#endif
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/SecurityModeControl.cpp:60]: Space

File: lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.cpp [hunk 19]
```diff
@@ -42,25 +49,26 @@
 #include "lte/gateway/c/core/oai/lib/itti/intertask_interface.h"
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.cpp]: Move header files related to itti, bstring and hash libraries under extern C

File: lte/gateway/c/core/oai/tasks/nas/emm/emm_headers.hpp [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+        authentication
+        security mode control
+        identification
+        EMM information
+
+*****************************************************************************/
+#ifndef FILE_EMM_HEARDERS_SEEN
+#define FILE_EMM_HEADERS_SEEN
+
+/*TODO: This file has temporary function declarations to
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/emm_headers.hpp:17]: Add the issue id, https://github.com/magma/magma/issues/13096 to handle this

File: lte/gateway/c/core/oai/tasks/nas/emm/sap/emm_cn.cpp [hunk 21]
```diff
@@ -38,11 +38,18 @@
...
+#endif
+#include "lte/gateway/c/core/common/common_defs.h"
+#include "lte/gateway/c/core/oai/common/log.h"
 #include "lte/gateway/c/core/oai/common/common_types.h"
 #include "lte/gateway/c/core/oai/common/conversions.h"
-#include "lte/gateway/c/core/oai/common/log.h"
+#include "lte/gateway/c/core/oai/lib/bstr/bstrlib.h"
+#ifdef __cplusplus
+}
+#endif
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/sap/emm_cn.cpp:53]: space

File: lte/gateway/c/core/oai/tasks/nas/emm/sap/emm_fsm.cpp [hunk 22]
```diff
@@ -18,15 +18,23 @@
...
+extern "C" {
+#endif
 #include "lte/gateway/c/core/common/assertions.h"
 #include "lte/gateway/c/core/common/common_defs.h"
 #include "lte/gateway/c/core/oai/common/common_types.h"
 #include "lte/gateway/c/core/oai/common/log.h"
+#ifdef __cplusplus
+}
+#endif
 #include "lte/gateway/c/core/oai/include/mme_app_ue_context.h"
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/sap/emm_fsm.cpp:34]: space

File: lte/gateway/c/core/oai/tasks/nas/emm/sap/emm_send.cpp [hunk 23]
```diff
@@ -20,23 +20,30 @@
...
+#ifdef __cplusplus
+}
+#endif
 #include "lte/gateway/c/core/oai/include/mme_config.h"
 #include "lte/gateway/c/core/oai/include/nas/securityDef.h"
 #include "lte/gateway/c/core/oai/lib/3gpp/3gpp_23.003.h"
 #include "lte/gateway/c/core/oai/lib/3gpp/3gpp_24.008.h"
-#include "lte/gateway/c/core/oai/lib/3gpp/3gpp_24.301.h"
 #include "lte/gateway/c/core/oai/lib/3gpp/3gpp_36.401.h"
 #include "lte/gateway/c/core/oai/lib/bstr/bstrlib.h"
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/sap/emm_send.cpp]: Move under extern C

File: lte/gateway/c/core/oai/tasks/nas/nas_network.cpp [hunk 24]
```diff
@@ -15,11 +15,18 @@
...
  */
 
+#ifdef __cplusplus
+extern "C" {
+#endif
 #include "lte/gateway/c/core/oai/common/log.h"
+#ifdef __cplusplus
+}
+#endif
 #include "lte/gateway/c/core/oai/include/mme_config.h"
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/nas_network.cpp:26]: space

File: lte/gateway/c/core/oai/tasks/nas/emm/emm_headers.hpp [hunk 25]
```diff
@@ -0,0 +1,245 @@
+/*
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas/emm/emm_headers.hpp:1]: Change the copyright

**Review by rsarwad — APPROVED** [review]
LGTM

**Pull request merged**